op-e2e/actions/proofs/isthmus_setcode_tx_test.go

```diff
@@ -0,0 +1,142 @@
+package proofs_test
+
+import (
+	"bytes"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/core/vm/program"
+	"github.com/holiman/uint256"
+	"github.com/stretchr/testify/require"
+
+	batcherFlags "github.com/ethereum-optimism/optimism/op-batcher/flags"
+	actionsHelpers "github.com/ethereum-optimism/optimism/op-e2e/actions/helpers"
+	"github.com/ethereum-optimism/optimism/op-e2e/actions/proofs/helpers"
+)
+
+func TestSetCodeTxTypeIsthmus(gt *testing.T) {
+	t := actionsHelpers.NewDefaultTesting(gt)
+
+	var (
+		aa = common.HexToAddress("0x000000000000000000000000000000000000aaaa")
+		bb = common.HexToAddress("0x000000000000000000000000000000000000bbbb")
+	)
+
+	// hardcoded because it's not available until after we need it
+	bobAddr := common.HexToAddress("0x14dC79964da2C08b23698B3D3cc7Ca32193d9955")
+
+	// Create 2 contracts, (1) writes 42 to slot 42, (2) calls (1)
+	store42Program := program.New().Sstore(0x42, 0x42)
+	callBobProgram := program.New().Call(nil, bobAddr, 1, 0, 0, 0, 0)
+
+	alloc := *actionsHelpers.DefaultAlloc
+	alloc.L2Alloc = make(map[common.Address]types.Account)
+	alloc.L2Alloc[aa] = types.Account{
+		Code: store42Program.Bytes(),
+	}
+	alloc.L2Alloc[bb] = types.Account{
+		Code: callBobProgram.Bytes(),
+	}
+
+	testCfg := &helpers.TestCfg[interface{}]{
+		Hardfork: helpers.Isthmus,
+		Allocs:   &alloc,
+	}
+
+	tp := helpers.NewTestParams()
+	env := helpers.NewL2FaultProofEnv(t, testCfg, tp, &actionsHelpers.BatcherCfg{
+		MinL1TxSize:          0,
+		MaxL1TxSize:          128_000,
+		DataAvailabilityType: batcherFlags.CalldataType,
+	})
+
+	require.Equal(gt, env.Bob.Address(), bobAddr)
+
+	cl := env.Engine.EthClient()
+
+	env.Sequencer.ActL2PipelineFull(t)
+	env.Miner.ActEmptyBlock(t)
+	env.Sequencer.ActL2StartBlock(t)
+
+	aliceSecret := env.Alice.L2.Secret()
+	bobSecret := env.Bob.L2.Secret()
+
+	chainID := env.Sequencer.RollupCfg.L2ChainID
+
+	// Sign authorization tuples.
+	// The way the auths are combined, it becomes
+	// 1. tx -> addr1 which is delegated to 0xaaaa
+	// 2. addr1:0xaaaa calls into addr2:0xbbbb
+	// 3. addr2:0xbbbb  writes to storage
+	auth1, err := types.SignSetCode(aliceSecret, types.SetCodeAuthorization{
+		ChainID: *uint256.MustFromBig(chainID),
+		Address: bb,
+		Nonce:   1,
+	})
+	require.NoError(gt, err, "failed to sign auth1")
+	auth2, err := types.SignSetCode(bobSecret, types.SetCodeAuthorization{
+		Address: aa,
+		Nonce:   0,
+	})
+	require.NoError(gt, err, "failed to sign auth2")
+
+	txdata := &types.SetCodeTx{
+		ChainID:   uint256.MustFromBig(chainID),
+		Nonce:     0,
+		To:        env.Alice.Address(),
+		Gas:       500000,
+		GasFeeCap: uint256.NewInt(5000000000),
+		GasTipCap: uint256.NewInt(2),
+		AuthList:  []types.SetCodeAuthorization{auth1, auth2},
+	}
+	signer := types.NewIsthmusSigner(chainID)
+	tx := types.MustSignNewTx(aliceSecret, signer, txdata)
```

Comment:
`env.Alice` is a test user helper that already provides some functionality to generate transactions and send them, see method `BasicUser.ActMakeTx` and other methods that modify the tx that is generated here. It may be a low lift to add SetCode transactions to the basic user to create these more easily in the future, e.g. as a `ActMakeSetCodeTx`. Optional, but nice if it's not too much work. We could then also use it in the other test about dropping pre-Isthmus batches with SetCode txs.

Reply:
I think this is a good idea, but the current method generates a random tx, and this requires a genesis alloc to exist to work properly. I could use the signer returned here, but we need the secret anyway for signing the SetCodeAuthorization.

The authorization might definitely warrant a new method though since I don't think the signer can sign authorizations (which is why I needed the new `SecretKey()` method).